node/data_node.go

```diff
@@ -0,0 +1,341 @@
+// Copyright (C) 2019-2023 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+// Package node is the Algorand node itself, with functions exposed to the frontend
+package node
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/algorand/go-algorand/agreement"
+	"github.com/algorand/go-algorand/catchup"
+	"github.com/algorand/go-algorand/config"
+	"github.com/algorand/go-algorand/data"
+	"github.com/algorand/go-algorand/data/account"
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/data/bookkeeping"
+	"github.com/algorand/go-algorand/data/transactions"
+	"github.com/algorand/go-algorand/ledger"
+	"github.com/algorand/go-algorand/ledger/ledgercore"
+	"github.com/algorand/go-algorand/logging"
+	"github.com/algorand/go-algorand/network"
+	"github.com/algorand/go-algorand/network/messagetracer"
+	"github.com/algorand/go-algorand/rpcs"
+	"github.com/algorand/go-algorand/stateproof"
+	"github.com/algorand/go-algorand/util/db"
+	"github.com/algorand/go-algorand/util/execpool"
+)
+
+// AlgorandDataNode implements sync mode/ledger delta APIs and disables participation-related methods
+type AlgorandDataNode struct {
+	AlgorandFullNode
+}
+
+// MakeData sets up an Algorand data node
+func MakeData(log logging.Logger, rootDir string, cfg config.Local, phonebookAddresses []string, genesis bookkeeping.Genesis) (*AlgorandDataNode, error) {
```

Comment:
I'm not sure which is better (or worse)... the copy/paste from `MakeFull` or having a bunch of `if dataMode { ... } else { ... }` conditions.

Is there a common subset that could be shared in a `makeNode` function? Comparing side-by-side it looks like `[]ledgercore.Blocklistener` could be an argument. I guess there would be some additional services that get added onto the result of `makeNode`.

Reply:
After talking with Eric we decided to keep going with the approach that he's started.

One side effect of that is that testing becomes more of a burden, because there are now multiple code paths doing similar things. Do we also need to duplicate the tests that cover things like fast catchup?